### PR TITLE
[WIP]Add beam search feature in transformer model

### PR DESCRIPTION
**Pull request by majing921201**
## What changes were proposed in this pull request?

Add beam search feature to support finding the translated sequence with the highest probability

## How was this patch tested?

UT

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BeamSearchTest.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.tensor._
+import com.intel.analytics.bigdl.numeric.NumericFloat
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+object BeamSearchTest {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BeamSearchTest.scala]: just for test? If so, please move it to `test` part

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor._
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+class SequenceBeamSearch[T: ClassTag](
+  val symbolsToLogitsFn: (Tensor[T], Tensor[Int], Map[String, Map[String, Tensor[T]]],
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: add comments to explain parameters

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+    var outputs = true
+    for (i <- 1 to sizeT(0)) {
+      outputs &&= ev.isGreater(tensor1.valueAt(i), tensor2.valueAt(i))
+    }
+    outputs
+  }
+
+  def continueSearch(state: Map[String, Any]): Boolean = {
+    val inf = 1.0 * 1e7 * (-1)
+    val i = state("CUR_INDEX").asInstanceOf[Tensor[Int]].clone()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: not use clone for each time

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+    val alive_log_probs = state("ALIVE_LOG_PROBS").asInstanceOf[Tensor[T]].clone()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: same

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+    val finished_flags = state("FINISHED_FLAGS").asInstanceOf[Tensor[Boolean]].clone()
+    var notAtMaxDecodeLength = true
+    if (i(Array(1)) < maxDecodeLength) {
+      notAtMaxDecodeLength = true
+    } else {
+      notAtMaxDecodeLength = false
+    }
+    val maxLengthNorm = lengthNormalization(alpha, maxDecodeLength)
+    val bestAliveScores = alive_log_probs.select(2, 1).clone() / maxLengthNorm
+    val newFinishedFlags = Tensor[T](finished_flags.size())
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: Not create new tensor for each time. You can create a global tensor and just update its values.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+    newState
+  }
+
+  def createInitialState(initialID: Tensor[T], initialCache:
+    Map[String, Map[String, Tensor[T]]] = null, initialEncoderDecoder:
+    Map[String, Tensor[T]] = null): Map[String, Any] = {
+    val curIndex = Tensor(Array(0), Array(1))
+    var aliveSeq = extendBeamSize(initialID, beamSize)
+    aliveSeq = expandDim(aliveSeq, 2)
+    val inf = 1.0f * 1e7f * (-1)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: it should be a global constant

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+    }
+    val finishedState = state
+    val aliveSeq = finishedState("ALIVE_SEQ").asInstanceOf[Tensor[T]].clone()
+    val aliveLogProbs = finishedState("ALIVE_LOG_PROBS").asInstanceOf[Tensor[T]].clone()
+    var finishedSeq = finishedState("FINISHED_SEQ").asInstanceOf[Tensor[T]].clone()
+    var finishedScores = finishedState("FINISHED_SCORES").asInstanceOf[Tensor[T]].clone()
+    val finishedFlags = finishedState("FINISHED_FLAGS").asInstanceOf[Tensor[Boolean]].clone()
+    finishedSeq = where(reduceAny(finishedFlags), finishedSeq, aliveSeq)
+    finishedScores = where(reduceAny(finishedFlags), finishedScores, aliveLogProbs)
+    println(finishedSeq)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: delete println

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+    val tileDim = new Array[Int](tensor.dim()).map(a => a + 1)
+    tileDim(1) = beamSize
+    tensor.repeatTensor(tileDim)
+  }
+
+  def lengthNormalization(alpha: Float, length: Int): T = {
+    ev.pow(ev.fromType[Double](5.0 + length / 6.0), ev.fromType[Float](alpha))
+  }
+
+  def boolToFloat(b: Boolean): T = {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: define method as private for internal fuc.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+    val shape = logits.size()
+    val getExp = Tensor[T](shape)
+    getExp.applyFun[T](logits, x => ev.exp(x))
+    val getSumExp = getExp.sum(3)
+    val getLogSumExp = Tensor[T](getSumExp.size())
+    getLogSumExp.applyFun[T](getSumExp, x => ev.log(x))
+    logits - getLogSumExp.repeatTensor(Array(1, 1, shape(2)))
+  }
+
+  def gatherNd(tensor: Tensor[T], indices: Tensor[T]): Tensor[T] = {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: Add comments for main method.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+        for (j <- 1 to shape2(1)) {
+          slices((i - 1) * shape2(1) + j - 1) = tensor.valueAt(ev.toType[Int](ev.plus
+          (indices.valueAt(i, j, 1), ev.fromType[Float](1.0f))), ev.toType[Int]
+            (ev.plus(indices.valueAt(i, j, 2), ev.fromType[Float](1.0f))))
+        }
+      }
+    }
+    shape1(0) = shape2(0)
+    shape1(1) = shape2(1)
+    Tensor(slices, shape1)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: Same. Not create new tensor for each time

**Comment by zhangxiaoli73** [comment]
You need add more tests to compare your results with tf

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+ * @param maxDecodeLength maximum length to decoded sequence
+ * @param eosID id of eos token, used to determine when a sequence has finished
+ * @param initialID Starting ids for each batch item.
+ * @param numHiddenLayers number of hidden layers
+ * @param hiddenSize size of hidden layer
+ * @param encoderOutputs outputs after encoding
+ * @param encoderDecoderAttentionBias bias for encoder decoder attention layer.
+ */
+class SequenceBeamSearch[T: ClassTag](
+  val symbolsToLogitsFn: (Tensor[T], Tensor[Int], Tensor[T], Tensor[T], List[Tensor[T]],
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: change to a tensor input

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 12]
```diff
@@ -0,0 +1,562 @@
...
+    List[Tensor[T]]) => (Tensor[T], Tensor[T], Tensor[T], List[Tensor[T]], List[Tensor[T]]),
+  val vocabSize: Int,
+  val batchSize: Int,
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: get from encoder output

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 13]
```diff
@@ -0,0 +1,562 @@
...
+  val beamSize: Int,
+  val alpha: Float,
+  val maxDecodeLength: Int,
+  val eosID: Float,
+  val initialID: Tensor[T],
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: no need

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 14]
```diff
@@ -0,0 +1,562 @@
...
+    tileDim(1) = beamSize
+    tensor.repeatTensor(tileDim)
+  }
+
+  private def lengthNormalization(alpha: Float, length: Int): T = {
+    ev.pow(ev.fromType[Double](5.0 + length / 6.0), ev.fromType[Float](alpha))
+  }
+
+  private def boolToFloat(b: Boolean): T = {
+    if (b) ev.fromType[Float](1.0f)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: ev.one

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearchSpec.scala [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    val vocab_size = 4
+    val decode_length = 10
+    val alpha: Float = 0.0f
+    val eosId = 1.0f
+    val numHiddenLayers = 2
+    val hiddenSize = 5
+    val inputLength = 6
+    val beamSearch = new SequenceBeamSearch[Float](vocab_size, batch_size,
+     beam_size, alpha, decode_length, eosId, numHiddenLayers, hiddenSize)
+    val encodeOutputs = Tensor[Float](batch_size, inputLength, hiddenSize)
```
Comment (zhangxiaoli73) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearchSpec.scala]: not use zero input

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearchSpec.scala [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    val encoderDecoderAttentionBias = Tensor[Float](batch_size, 1, 1, inputLength)
```
Comment (zhangxiaoli73) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearchSpec.scala]: not use zero input

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 17]
```diff
@@ -0,0 +1,568 @@
...
+
+  // Gather top beams from nested structure.
+  private def gatherTopkBeams(tensor: Tensor[T], scoreOrLogProb: Tensor[T],
+    batchSize: Int, beamSize: Int): Tensor[T] = {
+    val (_, topkIndexes) = scoreOrLogProb.topk(beamSize, -1, false)
+    topkIndexes.apply1(e => ev.minus(e, ev.fromType[Float](1.0f)))
+    gatherBeams(tensor, topkIndexes, batchSize, beamSize)
+  }
+
+  def getLogitFn(): SequenceBeamSearch[T] = {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: setLogitFn

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 18]
```diff
@@ -0,0 +1,568 @@
...
+  private val aliveLogProbs = Tensor[T](batchSize, beamSize)
+  private var finishedSeq = Tensor[T]
+  private val aliveSeq = Tensor[T]
+  private val finishedFlags = Tensor[Boolean](batchSize, beamSize)
+  private val finishedFlagsSeq = Tensor[T](batchSize, beamSize * 2)
+  private var finishedScores = Tensor[T](batchSize, beamSize)
+  private val gatherTensor = Tensor[T]
+  private val topkSeq = Tensor[T]
+  private val topkLogProbs = Tensor[T]
+  private val topkScore = Tensor[T]
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: add `clearState` to clear internal tensor

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 19]
```diff
@@ -0,0 +1,568 @@
...
+            a.setValue(i + 1, j, b.valueAt(i + 1, j))
+          }
+        }
+      }
+    }
+    a
+  }
+
+  override def updateOutput(input: Table): Activity = {
+    val encoderOutputs = input[Tensor[T]](1)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: Check whether `symbolsToLogits` is set

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 20]
```diff
@@ -0,0 +1,594 @@
...
+
+  override def updateGradInput(input: Table, gradOutput: Activity): Table = {
+    gradInput = gradOutput.toTable
+    gradInput
+  }
+
+  override def clearState(): this.type = {
+    super.clearState()
+    batchSize = 0
+    inf = 0.0f
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: not change

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 21]
```diff
@@ -0,0 +1,594 @@
...
+  val vocabSize: Int,
+  val beamSize: Int,
+  val alpha: Float,
+  val maxDecodeLength: Int,
+  val eosID: Float,
+  val numHiddenLayers: Int,
+  val hiddenSize: Int)(implicit ev: TensorNumeric[T])
+  extends AbstractModule[Table, Activity, T] {
+
+  private var inf = 0.0f
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: not 0.0

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 22]
```diff
@@ -0,0 +1,594 @@
...
+    aliveSeq.set()
+    finishedFlags.set()
+    finishedFlagsSeq.set()
+    finishedScores.set()
+    gatherTensor.set()
+    topkSeq.set()
+    topkLogProbs.set()
+    topkScore.set()
+    topkFlags.set()
+    symbolToLogits = null
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: not clear

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearchSpec.scala [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+        T(T(0.0, 1.0, 0.0, 0.0, 0.0),
+        T(0.0, 0.0, 1.0, 0.0, 0.0),
+        T(0.0, 2.0, 1.0, 0.0, 0.0))))
+    val expectedOutputScore = Tensor[Float](
+      T(T(-1.2615868, -2.2131736, -3.1647604),
+        T(-1.3734006, -2.4668012, -2.715382)))
+    outputSeq should be(expectedOutputSeq)
+    outputScore should be(expectedOutputScore)
+  }
+}
```
Comment (zhangxiaoli73) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearchSpec.scala]: SerialTest

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearchSpec.scala [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+
+class SequenceBeamSearchSerialTest extends ModuleSerializationTest{
```
Comment (majing921201) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearchSpec.scala]: add SerialTest

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala [hunk 25]
```diff
@@ -0,0 +1,585 @@
...
+    output
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Activity): Table = {
+    gradInput = gradOutput.toTable
+    gradInput
+  }
+
+  override def clearState(): this.type = {
+    super.clearState()
```
Comment (majing921201) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala]: remove inf and symbolToLogits function clear

**Pull request merged**